src/util.h

```diff
@@ -252,4 +252,45 @@ SECP256K1_GNUC_EXT typedef unsigned __int128 uint128_t;
 SECP256K1_GNUC_EXT typedef __int128 int128_t;
 #endif
 
+#ifndef __has_builtin
+#define __has_builtin(x) 0
+#endif
+
+static SECP256K1_INLINE int secp256k1_ctz32_var(uint32_t x) {
+#if (__has_builtin(__builtin_ctz) || SECP256K1_GNUC_PREREQ(3,4))
+    if (((unsigned)UINT32_MAX) == UINT32_MAX) {
```

Comment:
It may be simpler and more correct to use  `__builtin_types_compatible_p` here to check for equality of types, which existed before `__builtin_ctz` and friends.

https://gcc.gnu.org/onlinedocs/gcc-3.1/gcc/Other-Builtins.html

Reply:
In practice that would work, but it's not exactly what you want: if `unsigned` happens to be larger than 32 bits, we still want this branch to be taken.

Reply:
Oh! I understand now. 

My point was that this currently checks only one direction, namely if `unsigned` is wider or equal to `uint32_t` (and not the other way around). But I see that this is exactly what we want to check, at least if `x` is non-zero. Neat.

Can you add a comment to this conditional? There should also be a comment that explains the semantics of the function (including that the result is undefined if `x` is 0, which is the case even for `__builtin_ctz`). 

(I know this is a draft PR, we can postpone all this to a more stable version.)

Reply:
Done.